--- Day 9: Mirage Maintenance ---

You ride the camel through the sandstorm and stop where the ghost's maps told you to stop. The sandstorm subsequently subsides, somehow seeing you standing at an oasis!

The camel goes to get some water and you stretch your neck. As you look up, you discover what must be yet another giant floating island, this one made of metal! That must be where the parts to fix the sand machines come from.

There's even a hang glider partially buried in the sand here; once the sun rises and heats up the sand, you might be able to use the glider and the hot air to get all the way up to the metal island!

While you wait for the sun to rise, you admire the oasis hidden here in the middle of Desert Island. It must have a delicate ecosystem; you might as well take some ecological readings while you wait. Maybe you can report any environmental instabilities you find to someone so the oasis can be around for the next sandstorm-worn traveler.

You pull out your handy Oasis And Sand Instability Sensor and analyze your surroundings. The OASIS produces a report of many values and how they are changing over time (your puzzle input). Each line in the report contains the history of a single value. For example:

0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45

To best protect the oasis, your environmental report should include a prediction of the next value in each history. To do this, start by making a new sequence from the difference at each step of your history. If that sequence is not all zeroes, repeat this process, using the sequence you just generated as the input sequence. Once all of the values in your latest sequence are zeroes, you can extrapolate what the next value of the original history should be.

In the above dataset, the first history is 0 3 6 9 12 15. Because the values increase by 3 each step, the first sequence of differences that you generate will be 3 3 3 3 3. Note that this sequence has one fewer value than the input sequence because at each step it considers two numbers from the input. Since these values aren't all zero, repeat the process: the values differ by 0 at each step, so the next sequence is 0 0 0 0. This means you have enough information to extrapolate the history! Visually, these sequences can be arranged like this:

0   3   6   9  12  15
  3   3   3   3   3
    0   0   0   0

To extrapolate, start by adding a new zero to the end of your list of zeroes; because the zeroes represent differences between the two values above them, this also means there is now a placeholder in every sequence above it:

0   3   6   9  12  15   B
  3   3   3   3   3   A
    0   0   0   0   0

You can then start filling in placeholders from the bottom up. A needs to be the result of increasing 3 (the value to its left) by 0 (the value below it); this means A must be 3:

0   3   6   9  12  15   B
  3   3   3   3   3   3
    0   0   0   0   0

Finally, you can fill in B, which needs to be the result of increasing 15 (the value to its left) by 3 (the value below it), or 18:

0   3   6   9  12  15  18
  3   3   3   3   3   3
    0   0   0   0   0

So, the next value of the first history is 18.

Finding all-zero differences for the second history requires an additional sequence:

1   3   6  10  15  21
  2   3   4   5   6
    1   1   1   1
      0   0   0

Then, following the same process as before, work out the next value in each sequence from the bottom up:

1   3   6  10  15  21  28
  2   3   4   5   6   7
    1   1   1   1   1
      0   0   0   0

So, the next value of the second history is 28.

The third history requires even more sequences, but its next value can be found the same way:

10  13  16  21  30  45  68
   3   3   5   9  15  23
     0   2   4   6   8
       2   2   2   2
         0   0   0

So, the next value of the third history is 68.

If you find the next value for each history in this example and add them together, you get 114.

Analyze your OASIS report and extrapolate the next value for each history. What is the sum of these extrapolated values?


In [ ]:
class History:
    def __init__(self, input):
        self.steps = self.__create(input)
        self.extrapolation = self.extrapolate()

    def __create(self, input):
        steps = []
        step_values = [int(x) for x in input.split()]

        all_zeroes = False
        row = 0
        while not all_zeroes:
            if row == 0:
                steps.append(step_values)
            else:
                previous_steps = steps[row - 1]
                current_steps = []
                for i in range(1, len(previous_steps)):
                    current_steps.append(previous_steps[i] - previous_steps[i - 1])

                steps.append(current_steps)
        
                all_zeroes = all([x == 0 for x in current_steps])
            row += 1
        
        return steps        

    def __str__(self):
        value = ""
        for index, step in enumerate(self.steps):
            tabs_string = " " * index
            step_string = " ".join([str(x) for x in step])
            value += f"{tabs_string}{step_string}\n"

        value += f"extrapolation: {str(self.extrapolation)}"
        return value

    def extrapolate(self):
        last_step_value = 0
        for index, step in enumerate(reversed(self.steps)):
            if index == 0:
                last_step_value = step[-1]
                continue   
            else:
                step_value = step[-1]
                last_step_value = step_value + last_step_value 
        return last_step_value
            

input1 = """0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45"""

with open('input.txt', 'r') as file:
    input3 = file.read()

inputs = input3.split("\n")
histories = []
for input in inputs:
    history = History(input)
    print(history)
    print()
    histories.append(history)

sum = 0
for history in histories:
    sum += history.extrapolation

print(sum)

Your puzzle answer was 1974913025.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---

Of course, it would be nice to have even more history included in your report. Surely it's safe to just extrapolate backwards as well, right?

For each history, repeat the process of finding differences until the sequence of differences is entirely zero. Then, rather than adding a zero to the end and filling in the next values of each previous sequence, you should instead add a zero to the beginning of your sequence of zeroes, then fill in new first values for each previous sequence.

In particular, here is what the third example history looks like when extrapolating back in time:

5  10  13  16  21  30  45
  5   3   3   5   9  15
   -2   0   2   4   6
      2   2   2   2
        0   0   0

Adding the new values on the left side of each sequence from bottom to top eventually reveals the new left-most history value: 5.

Doing this for the remaining example data above results in previous values of -3 for the first history and 0 for the second history. Adding all three new values together produces 2.

Analyze your OASIS report again, this time extrapolating the previous value for each history. What is the sum of these extrapolated values?


In [77]:
! pip install termcolor

from termcolor import colored


[notice] A new release of pip is available: 23.2.1 -> 23.3.1
[notice] To update, run: C:\Users\dkwar\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [88]:
class History:
    def __init__(self, input):
        self.steps = self.__create(input)
        self.front_extrapolation, self.front_extrapolation_values = self.extrapolate_front()
        self.back_extrapolation, self.back_extrapolation_values = self.extrapolate_back()
   
    def extrapolate_front(self):
        previous_step_value = 0
        values = []
        for step in reversed(self.steps):
            new_value = step[-1] + previous_step_value

            values.append(new_value) 

            previous_step_value = new_value
        return new_value, values
    
    def extrapolate_back(self):
        previous_step_value = 0
        values = []
        for step in reversed(self.steps):
            new_value = step[0] - previous_step_value

            values.append(new_value) 

            previous_step_value = new_value
        return new_value, values

    def __create(self, input):
        steps = []
        step_values = [int(x) for x in input.split()]

        all_zeroes = False
        row = 0
        while not all_zeroes:
            if row == 0:
                steps.append(step_values)
            else:
                previous_steps = steps[row - 1]
                current_steps = []
                for i in range(1, len(previous_steps)):
                    current_steps.append(previous_steps[i] - previous_steps[i - 1])

                steps.append(current_steps)
        
                all_zeroes = all([x == 0 for x in current_steps])
            row += 1
        
        return steps       
    
    def __str__(self):
        value = ""
        max_row_length = max(len(" ".join(map(str, step))) for step in self.steps)
        back_values_reversed = list(reversed(self.back_extrapolation_values))
        front_values_reversed = list(reversed(self.front_extrapolation_values))

        for index, step in enumerate(self.steps):
            step_string = " ".join(map(str, step))
            back_text = colored(str(back_values_reversed[index]), "red") if index < len(back_values_reversed) else "  "
            front_text = colored(str(front_values_reversed[index]), "green") if index < len(front_values_reversed) else "  "

            value += f"{back_text} {step_string} {front_text}".center(max_row_length) + "\n"

        value += f"Front extrapolation: {self.front_extrapolation}\n"
        value += f"Back extrapolation: {self.back_extrapolation}\n"
        return value
            

text  = """5  10  13  16  21  30  45
  5   3   3   5   9  15
   -2   0   2   4   6
      2   2   2   2
        0   0   0"""

input1 = """0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45"""

with open('input.txt', 'r') as file:
    input3 = file.read()

inputs = input3.split("\n")
histories = []
for input in inputs:
    history = History(input)
    histories.append(history)

sum_extrapolate_front = 0
sum_extrapolate_back = 0
for history in histories:
    sum_extrapolate_front += history.front_extrapolation
    sum_extrapolate_back += history.back_extrapolation

print("Sum extrapolate front: ", sum_extrapolate_front, " Sum extrapolate back: ", sum_extrapolate_back)

for history in histories:
    print(history)
    print()

Sum extrapolate front:  1974913025  Sum extrapolate back:  884
15 23 45 81 131 195 273 365 471 591 725 873 1035 1211 1401 1605 1823 2055 2301 2561 2835 3123 3425
8 22 36 50 64 78 92 106 120 134 148 162 176 190 204 218 232 246 260 274 288 302
     14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14     
                0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0                 
Front extrapolation: 3425
Back extrapolation: 15


9 10 9 7 18 72 228 601 1407 3030 6115 11691 21328 37332 62982 102813 162949 251490 378957 558799 807966 1147552 1603512
1 -1 -2 11 54 156 373 806 1623 3085 5576 9637 16004 25650 39831 60136 88541 127467 179842 249167 339586 455960
-2 -1 13 43 102 217 433 817 1462 2491 4061 6367 9646 14181 20305 28405 38926 52375 69325 90419 116374
1 14 30 59 115 216 384 645 1029 1570 2306 3279 4535 6124 8100 10521 13449 16950 21094 25955
      13 16 29 56 101 168 261 384 541 736 973 1256 1589 1976 2421 2928 3501 4144 4861      
              3 13 27 45 67 93 123 157 195 2

Your puzzle answer was 884.

Both parts of this puzzle are complete! They provide two gold stars: **